dhis-2/dhis-e2e-test/src/main/java/org/hisp/dhis/actions/SystemActions.java

```diff
@@ -46,30 +51,52 @@ public SystemActions()
         super( "/system" );
     }
 
+    public ApiResponse getTask( String taskType, String taskId )
+    {
+        return get( "/tasks/" + taskType + "/" + taskId );
+    }
+
+    /**
+     * Waits until the task is completed and returns a response
+     *
+     * @param taskType
+     * @param taskId
+     * @return
+     */
     public ApiResponse waitUntilTaskCompleted( String taskType, String taskId )
     {
-        logger.info( "Waiting until task " + taskType + " with id " + taskId + "is completed" );
-        ApiResponse response = null;
-        boolean completed = false;
-        while ( !completed )
-        {
-            response = get( "/tasks/" + taskType + "/" + taskId );
-            response.validate().statusCode( 200 );
-            completed = response.extractList( "completed" ).contains( true );
-        }
+        Callable<Boolean> taskIsCompleted = () -> getTask( taskType, taskId )
+            .validateStatus( 200 )
+            .extractList( "completed" ).contains( true );
+
+        with()
+            .pollInterval( 1, TimeUnit.SECONDS )
+            .atMost( 20, TimeUnit.SECONDS )
+            .await().until( () -> taskIsCompleted.call() );
```

Comment:
not sure how this looks in the logs if the test failed. adding a message to `await()` https://github.com/awaitility/awaitility/wiki/Usage#advanced could be helpful

Reply:
I tried that, but it adds way too much noise and no useful output.